File: app/src/main/java/crux/bphc/cms/fragments/FolderModuleFragment.java
package crux.bphc.cms.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import crux.bphc.cms.R;
import crux.bphc.cms.interfaces.ClickListener;
import crux.bphc.cms.core.FileManager;
import crux.bphc.cms.models.course.Content;
import crux.bphc.cms.models.course.Module;
import crux.bphc.cms.utils.FileUtils;
import crux.bphc.cms.widgets.PropertiesAlertDialog;
import io.realm.Realm;
import io.realm.RealmList;
import kotlin.Unit;

public class FolderModuleFragment extends Fragment {

    private static final String MODULE_ID_KEY = "moduleID";
    private static final String COURSE_NAME_KEY = "courseName";

    Realm realm;

    private int MODULE_INSTANCE = 0;
    private String COURSE_NAME = "";

    private Module module;
    private RealmList<Content> contents;

    private FolderModuleFragment.FolderModuleAdapter mAdapter;
    private FileManager mFileManager;

    private MoreOptionsFragment.OptionsViewModel moreOptionsViewModel;

    public FolderModuleFragment() {

    }

    public static FolderModuleFragment newInstance(int moduleId, String courseName) {
        FolderModuleFragment fragment = new FolderModuleFragment();
        Bundle args = new Bundle();
        args.putInt(MODULE_ID_KEY, moduleId);
        args.putString(COURSE_NAME_KEY, courseName);
        fragment.setArguments(args);
        return fragment;
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            MODULE_INSTANCE = getArguments().getInt(MODULE_ID_KEY);
            COURSE_NAME = getArguments().getString(COURSE_NAME_KEY);
        }

        mFileManager = new FileManager(requireActivity(), COURSE_NAME, fileName -> {
            onDownloadComplete(fileName);
            return Unit.INSTANCE;
        });
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        realm = Realm.getDefaultInstance();
        return inflater.inflate(R.layout.fragment_folder_module, container, false);
    }

    @Override
    public void onViewCreated(@NotNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        moreOptionsViewModel = new ViewModelProvider(requireActivity()).get(MoreOptionsFragment.OptionsViewModel.class);

        ClickListener mClickListener = (object, position) -> {
            Content content = (Content) object;
            downloadOrOpenFile(content, false);
            return true;
        };

        module = realm.where(Module.class).equalTo("instance", MODULE_INSTANCE).findFirst();
        contents = module != null ? module.getContents() : null;

        RecyclerView mRecyclerView = view.findViewById(R.id.files);
        final LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        mRecyclerView.setLayoutManager(layoutManager);

        mAdapter = new FolderModuleFragment.FolderModuleAdapter(mClickListener, new ArrayList<>());
        mRecyclerView.setAdapter(mAdapter);

        updateContents();
    }

    private void updateContents() {
        mAdapter.setContents(contents);
    }

    private void downloadOrOpenFile(Content content, boolean forceDownload) {
        if (forceDownload || !mFileManager.isModuleContentDownloaded(content)) {
            Toast.makeText(getActivity(), "Downloading file - " + content.getFileName(), Toast.LENGTH_SHORT).show();
            mFileManager.downloadModuleContent(content, module);
        } else {
            mFileManager.openModuleContent(content);
        }
    }

    private void onDownloadComplete(String fileName) {
        mAdapter.notifyDataSetChanged();
        Content content;
        if (contents != null
                && (content = contents.where().equalTo("fileName", fileName).findFirst()) != null) {
            mFileManager.openModuleContent(content);
        }

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        realm.close();
    }

    private class FolderModuleAdapter extends RecyclerView.Adapter<FolderModuleFragment.FolderModuleAdapter.FolderModuleViewHolder> {

        private final List<Content> mContents;
        private final ClickListener mClickListener;

        public FolderModuleAdapter(ClickListener clickListener, List<Content> contents) {
            mClickListener = clickListener;
            mContents = contents;
        }

        public void setContents(List<Content> contents) {
            mContents.addAll(contents);
            notifyDataSetChanged();
        }

        @NotNull
        @Override
        public FolderModuleFragment.FolderModuleAdapter.FolderModuleViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            return new FolderModuleFragment.FolderModuleAdapter.FolderModuleViewHolder(inflater.inflate(R.layout.row_folder_module_file, parent, false));
        }

        @Override
        public void onBindViewHolder(FolderModuleFragment.FolderModuleAdapter.FolderModuleViewHolder holder, int position) {
            Content content = mContents.get(position);
            holder.bind(content);
        }

        @Override
        public int getItemCount() {
            return mContents.size();
        }


        public class FolderModuleViewHolder extends RecyclerView.ViewHolder {

            private final TextView fileName;
            private final ImageView fileIcon;
            private final ImageView download;
            private final ImageView ellipsis;
            private final LinearLayout clickWrapper;
            
            public FolderModuleViewHolder(View itemView) {
                super(itemView);

                fileName = itemView.findViewById(R.id.name);
                fileIcon = itemView.findViewById(R.id.icon);
                download = itemView.findViewById(R.id.download);
                ellipsis = itemView.findViewById(R.id.more);
                clickWrapper = itemView.findViewById(R.id.click_wrapper);
            }

            public void bind(Content content) {
                fileName.setText(content.getFileName());

                int icon = FileUtils.INSTANCE.getDrawableIconFromFileName(content.getFileName());
                if (icon != -1) {
                    fileIcon.setImageResource(icon);
                } else {
                    fileIcon.setImageResource(R.drawable.quiz); // Quiz because that's a question mark xD
                }

                boolean downloaded = mFileManager.isModuleContentDownloaded(content);
                if (downloaded) {
                    download.setImageResource(R.drawable.eye);
                }

                clickWrapper.setOnClickListener(view -> {
                    if (mClickListener != null) {
                        mClickListener.onClick(contents.get(getLayoutPosition()), getLayoutPosition());
                    }
                });

                ellipsis.setOnClickListener(view -> {
                    ArrayList<MoreOptionsFragment.Option> options = new ArrayList<>();
                    Observer<MoreOptionsFragment.Option> observer;  // to handle the selection

                    if (downloaded) {
                        options.addAll(Arrays.asList(
                                new MoreOptionsFragment.Option(0, "View", R.drawable.eye),
                                new MoreOptionsFragment.Option(1, "Re-Download", R.drawable.download),
                                new MoreOptionsFragment.Option(2, "Share", R.drawable.ic_share),
                                new MoreOptionsFragment.Option(3, "Properties", R.drawable.ic_info)
                        ));

                        observer = option -> {
                            if (option == null) return;
                            switch (option.getId()) {
                                case 0:
                                    downloadOrOpenFile(content, false);
                                    break;
                                case 1:
                                    downloadOrOpenFile(content, true);
                                    break;
                                case 2:
                                    mFileManager.shareModuleContent(content);
                                    break;
                                case 3:
                                    new PropertiesAlertDialog(getContext(), content).show();
                            }
                            moreOptionsViewModel.getSelection().removeObservers(requireActivity());
                            moreOptionsViewModel.clearSelection();
                        };
                    } else {
                        options.addAll(Arrays.asList(
                                new MoreOptionsFragment.Option(0, "Download", R.drawable.download),
                                new MoreOptionsFragment.Option(1, "Properties", R.drawable.ic_info)
                        ));

                        observer = option -> {
                            if (option == null) return;
                            switch (option.getId()) {
                                case 0:
                                    downloadOrOpenFile(content, false);
                                    break;
                                case 1:
                                    new PropertiesAlertDialog(getContext(), content).show();
                                    break;
                            }
                        };
                        moreOptionsViewModel.getSelection().removeObservers(requireActivity());
                        moreOptionsViewModel.clearSelection();
                    }
                    MoreOptionsFragment fragment = MoreOptionsFragment.newInstance(content.getFileName(), options);
                    fragment.show(requireActivity().getSupportFragmentManager(),
                            fragment.getTag());
                    moreOptionsViewModel.getSelection().observe(requireActivity(), observer);
                });
            }
        }
    }
}

File: app/src/main/java/crux/bphc/cms/fragments/FolderModuleFragment.kt
package crux.bphc.cms.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import crux.bphc.cms.R
import crux.bphc.cms.core.FileManager
import crux.bphc.cms.databinding.FragmentFolderModuleBinding
import crux.bphc.cms.databinding.RowFolderModuleFileBinding
import crux.bphc.cms.fragments.MoreOptionsFragment.OptionsViewModel
import crux.bphc.cms.interfaces.ClickListener
import crux.bphc.cms.models.course.Content
import crux.bphc.cms.models.course.Module
import crux.bphc.cms.utils.FileUtils
import crux.bphc.cms.widgets.PropertiesAlertDialog
import io.realm.Realm
import java.util.*

class FolderModuleFragment : Fragment() {
    private lateinit var realm: Realm
    private lateinit var mFileManager: FileManager
    private lateinit var moreOptionsViewModel: OptionsViewModel
    private lateinit var mAdapter: FolderModuleAdapter
    private lateinit var binding: FragmentFolderModuleBinding

    private var moduleInstance = 0
    private var courseName: String = ""
    private var module: Module? = null
    private var contents: List<Content> = emptyList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = FragmentFolderModuleBinding.inflate(layoutInflater)
        moduleInstance = requireArguments().getInt(MODULE_ID_KEY)
        courseName = requireArguments().getString(COURSE_NAME_KEY) ?: ""

        realm = Realm.getDefaultInstance()
        mFileManager = FileManager(requireActivity(), courseName) { fileName: String ->
            onDownloadComplete(fileName)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        moreOptionsViewModel = ViewModelProvider(requireActivity())[OptionsViewModel::class.java]

        val mClickListener = ClickListener { `object`: Any, _: Int ->
            val content = `object` as Content
            downloadOrOpenFile(content, false)
            true
        }

        val layoutManager = LinearLayoutManager(context)

        mAdapter = FolderModuleAdapter(mClickListener, ArrayList())
        binding.files.layoutManager = layoutManager
        binding.files.adapter = mAdapter

        module = realm.where(Module::class.java).equalTo("instance", moduleInstance).findFirst()
        contents = module?.contents ?: emptyList()
        updateContents()
    }

    private fun updateContents() {
        mAdapter.setContents(contents)
    }

    private fun downloadOrOpenFile(content: Content, forceDownload: Boolean) {
        if (forceDownload || !mFileManager.isModuleContentDownloaded(content)) {
            Toast.makeText(activity, "Downloading file - " + content.fileName, Toast.LENGTH_SHORT)
                .show()
            mFileManager.downloadModuleContent(content, module!!)
        } else {
            mFileManager.openModuleContent(content)
        }
    }

    private fun onDownloadComplete(fileName: String) {
        mAdapter.notifyDataSetChanged()
        val content = contents.stream().filter{ it.fileName == fileName }.findAny().orElse(null)
        content?.apply { mFileManager.openModuleContent(this) }
    }

    override fun onDestroy() {
        super.onDestroy()
        realm.close()
    }

    private inner class FolderModuleAdapter(
        private val mClickListener: ClickListener,
        private val contents: MutableList<Content>
    ) : RecyclerView.Adapter<FolderModuleAdapter.FolderModuleViewHolder>() {
        private val inflater: LayoutInflater = LayoutInflater.from(context)

        fun setContents(contents: List<Content>) {
            this.contents.addAll(contents)
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FolderModuleViewHolder {
            val itemBinding = RowFolderModuleFileBinding.inflate(inflater, parent, false)
            return FolderModuleViewHolder(itemBinding)
        }

        override fun onBindViewHolder(holder: FolderModuleViewHolder, position: Int) {
            holder.bind(contents[position])
        }

        override fun getItemCount(): Int {
            return contents.size
        }

        inner class FolderModuleViewHolder(val itemBinding: RowFolderModuleFileBinding) : RecyclerView.ViewHolder(itemBinding.root) {
            fun bind(content: Content) {
                itemBinding.name.text = content.fileName

                val icon = FileUtils.getDrawableIconFromFileName(content.fileName)
                itemBinding.icon.setImageResource(icon)

                val downloaded = mFileManager.isModuleContentDownloaded(content)
                if (downloaded) {
                    itemBinding.download.setImageResource(R.drawable.eye)
                }

                itemBinding.clickWrapper.setOnClickListener {
                    mClickListener.onClick(contents[layoutPosition], layoutPosition)
                }

                itemBinding.more.setOnClickListener {
                    val options = ArrayList<MoreOptionsFragment.Option>()
                    val observer: Observer<MoreOptionsFragment.Option?> // to handle the selection
                    if (downloaded) {
                        options.addAll(
                            listOf(
                                MoreOptionsFragment.Option(0, "View", R.drawable.eye),
                                MoreOptionsFragment.Option(1, "Re-Download", R.drawable.outline_file_download_24),
                                MoreOptionsFragment.Option(2, "Share", R.drawable.ic_share),
                                MoreOptionsFragment.Option(3, "Properties", R.drawable.ic_info)
                            )
                        )
                        observer = Observer { option: MoreOptionsFragment.Option? ->
                            if (option == null) return@Observer
                            when (option.id) {
                                0 -> downloadOrOpenFile(content, false)
                                1 -> downloadOrOpenFile(content, true)
                                2 -> mFileManager.shareModuleContent(content)
                                3 -> PropertiesAlertDialog(requireActivity(), content).show()
                            }
                            moreOptionsViewModel.selection.removeObservers(requireActivity())
                            moreOptionsViewModel.clearSelection()
                        }
                    } else {
                        options.addAll(
                            listOf(
                                MoreOptionsFragment.Option(0, "Download", R.drawable.outline_file_download_24),
                                MoreOptionsFragment.Option(1, "Properties", R.drawable.ic_info)
                            )
                        )
                        observer = Observer { option: MoreOptionsFragment.Option? ->
                            if (option == null) return@Observer
                            when (option.id) {
                                0 -> downloadOrOpenFile(content, false)
                                1 -> PropertiesAlertDialog(requireContext(), content).show()
                            }
                        }
                        moreOptionsViewModel.selection.removeObservers(requireActivity())
                        moreOptionsViewModel.clearSelection()
                    }

                    val fragment = MoreOptionsFragment.newInstance(content.fileName, options)
                    fragment.show(requireActivity().supportFragmentManager, fragment.tag)
                    moreOptionsViewModel.selection.observe(requireActivity(), observer)
                }
            }
        }
    }

    companion object {
        private const val MODULE_ID_KEY = "moduleID"
        private const val COURSE_NAME_KEY = "courseName"

        fun newInstance(moduleId: Int, courseName: String): FolderModuleFragment {
            val args = Bundle()
            args.putInt(MODULE_ID_KEY, moduleId)
            args.putString(COURSE_NAME_KEY, courseName)

            val fragment = FolderModuleFragment()
            fragment.arguments = args
            return fragment
        }
    }
}